docs/grid-introducer.rst

```diff
@@ -0,0 +1,104 @@
+.. -*- coding: utf-8 -*-
+
+Grid Introducer
+===============
+
+This document explains use and operation of the "Grid Introducer" system.
+This system replaces the earlier "Introducer node" system.
+
+The "Grid Introducer" system is a means for client nodes to initially learn about available storage nodes.
+It also allows client nodes to receive updates to this information.
+The goal of this document is to explain how to *use* this system,
+either as a user of a Tahoe-LAFS client node or as the operator of a Tahoe-LAFS grid or storage node.
+
+
+Use As A Client
+---------------
+
+You want your Tahoe-LAFS client to be able to find storage servers to access.
+This means your client needs both current and future connection information for those storage servers.
+By configuring your client with one small static connection hint,
+the "Grid Introducer" system enables the client to find more information and future updates to that information.
+
+Before attempting this configuration,
+you should have the Grid Introducer *cap* and *furl* which you received from the operator of a Tahoe-LAFS storage grid which uses the Grid Introducer system.
+
+The Grid Introducer is configured in ``tahoe.cfg`` in the ``[client]`` section's ``grid-introducer.cap`` and ``grid-introducer.furl`` items.
+For example::
+
+  [client]
+  grid-introducer.cap = URI:DIR2-RO:4bnx...
+  grid-introducer.furl = pb://sokl...@192.168.69.247:44801/eqpw...
+
+Start a Tahoe-LAFS client node with these items configured and the client will be able to find and follow all storage servers that are part of that grid.
+
```

Comment:
I wonder if the "connection hint" here should be "state" and not "config" because it should get updated (e.g. if that one storage-server is taken offline, this config will never be able to bootstrap again). If it's "state" you'd want to "add" and "remove" storage-servers from it and can use any/all of them to download new information. We already have a mechanism for this, AFAIK, since storage-server announcements are cached. Also "introducer-less" mode uses this.

That leaves "how to enroll a new client", because they can't just shove a thing into their config. Perhaps an "online" invite method (e.g. magic-wormhole) is too complex for this document? But it might be better to specify a single "invite a new client" document -- this *could* be communicated via magic-wormhole but also could simply be fed to a client via command-line (so that the magic-wormhole version is indeed simply "a built-in secure way to communicate that invite document").

So I'd be thinking along the lines of `tahoe join-grid <path-to-invite-document>` as the first version and `tahoe join-grid --code 3-foo-bar` or so for the improved magic-wormhole version. Then, all that stuff (the "cap" and a list of storage-servers) is put in "state" and the user doesn't copy-paste it. These are very long-term secrets and I don't think the user should really be trusted to not leak them (by having to put them into the config file).

The "invite document" is just the same information as above (although ideally with a list of server fURLs instead of just one).

Reply:
> I wonder if the "connection hint" here should be "state" and not "config" because it should get updated (e.g. if that one storage-server is taken offline, this config will never be able to bootstrap again). If it's "state" you'd want to "add" and "remove" storage-servers from it and can use any/all of them to download new information. We already have a mechanism for this, AFAIK, since storage-server announcements are cached. Also "introducer-less" mode uses this.

Oh.  Yea.  So basically what I wrote above about bootstrap spof and list vs singleton. :)  I'll put more words in the docs to make the extra intended client behavior clear.

Reply:
> That leaves "how to enroll a new client", because they can't just shove a thing into their config. Perhaps an "online" invite method (e.g. magic-wormhole) is too complex for this document? But it might be better to specify a single "invite a new client" document -- this could be communicated via magic-wormhole but also could simply be fed to a client via command-line (so that the magic-wormhole version is indeed simply "a built-in secure way to communicate that invite document").

I think they _can_ just shove a thing into their config.  The config is "what the user told me I could use to bootstrap" and the state is "everything I learned afterwards".  And the client behavior should be to synthesize a list of servers from these two pieces of information and try to retrieve announcements from each (all?) of those servers.

Reply:
> So I'd be thinking along the lines of tahoe join-grid <path-to-invite-document> as the first version and tahoe join-grid --code 3-foo-bar or so for the improved magic-wormhole version. Then, all that stuff (the "cap" and a list of storage-servers) is put in "state" and the user doesn't copy-paste it. These are very long-term secrets and I don't think the user should really be trusted to not leak them (by having to put them into the config file).

That seems more or less fine as a user experience, too, and could improve the bootstrap process - maybe even by sending _all_ of the current announcements.  I think there's probably value in a setup process that is independent of magic-wormhole.  So keeping either the tahoe.cfg route or the `tahoe join-grid <path>` thing even after there's a magic-wormhole-enabled version probably makes sense.

I'll try to edit some of this into the document too.

Reply:
Yeah, the general idea of the `tahoe join-grid <some file>` is that then it can be *exactly* the same for a magic-wormhole use-case *and* if it's a JSON blob, the user is very unlikely to copy/paste it wrong (because it's all or nothing).

So essentially this means the "grid-introducer" information is "some JSON/whatever document" and we have one delivery mechanism for now (out-of-band + command-line tool) and in the future we can have other delivery mechanisms (magic-wormhole).

Even without the CLI command, it could be that `tahoe.cfg` points at "a grid-introducer document/file" (which has the distinct advantage of not putting yet another secret in tahoe.cfg .. even if that ship kind-of sailed already with the introduer-furl).

Reply:
^ also makes it easier to evolve: if we change the information, it's still "a grid-introducer document" but the version might be 2 or 3 or whatever but we don't have to try and support some new `tahoe.cfg` items...

Reply:
Oh yea.  Separate file.  Great idea.  That definitely seems like the way to go.